clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -108,11 +108,15 @@ RawAddress CodeGenFunction::CreateTempAlloca(llvm::Type *Ty, CharUnits Align,
   if (AllocaAddr)
     *AllocaAddr = Alloca;
   llvm::Value *V = Alloca.getPointer();
+  assert((!getLangOpts().OpenCL ||
+          CGM.getTarget().getTargetAddressSpace(getASTAllocaAddressSpace()) ==
+              CGM.getTarget().getTargetAddressSpace(LangAS::opencl_private)) &&
+         "For OpenCL allocas must allocate in the private address space!");
```

Comment:
This assertion probably does not hold for nvptx. Alloca returns a generic pointer and is later lowered to the whatever they call private + cast in the backend

Reply:
Are you certain? That's not how I'm reading its [ASMAP](https://github.com/llvm/llvm-project/blob/39303e24b6f628f3c080f1b54bd12383a55b9b3a/clang/lib/Basic/Targets/NVPTX.h#L31), think this'd work just fine it aliases generic and private (and appears they designed their ASMap to handle precisely this case). Am I missing something in particular? The above is only a concern in Clang, once it's made it to the BE it's free to do whatever, really.

Reply:
That is the nature of the PTX hack. They pretend everything is 0, and then fix up stack later. The backend then runs NVPTXLowerAlloca to introduce a bunch of casts to show that it really was an addrspace(5) pointer to begin with around most use contexts

Reply:
Then this is safe / the assert won't flare and it matches existing behaviour, does it not?

Reply:
Yes (but will be invalid if they ever fix up that hack, as there have been posts suggesting favor of in the past)